### PR TITLE
#16 Dispatcher Retry & Dead Letter Queues

### PR DESCRIPTION
**Pull request by sbcd90**
@travis-minke-sap this is a solution proposal as per your approach 1.

### CODE REVIEW

File: pkg/dispatcher/dispatcher/retry.go [hunk 1]
```diff
@@ -58,13 +60,21 @@ func (d *Dispatcher) Dispatch(event *cloudevents.Event, subscription Subscriptio
...
 	if subscription.Delivery != nil && subscription.Delivery.DeadLetterSink != nil && !subscription.Delivery.DeadLetterSink.URI.IsEmpty() {
 		// TODO - Currently ignoring dead-letter configuration due to wrapping retry implementation - would send one deadletter for every retry :(
-		// deadLetterURL = subscription.Delivery.DeadLetterSink.URI.URL()
-		logger.Warn("Subscription Delivery DeadLetterSink Not Currently Supported!")
+		deadLetterURL = subscription.Delivery.DeadLetterSink.URI.URL()
+		// logger.Warn("Subscription Delivery DeadLetterSink Not Currently Supported!")
 	}
 
 	// Attempt To Dispatch The CloudEvent Message Via Knative Message Dispatcher With Retry Wrapper
 	err := retryRunner.Run(context.Background(), func(ctx context.Context) error {
```
Comment (lionelvillard) [inline pkg/dispatcher/dispatcher/retry.go:68]: FYI, the cloudevent SDK supports retries: https://github.com/cloudevents/sdk-go/blob/master/v2/protocol/http/protocol_retry.go#L16

Do you think you could leverage it?

File: pkg/dispatcher/dispatcher/retry.go [hunk 2]
```diff
@@ -58,13 +60,21 @@ func (d *Dispatcher) Dispatch(event *cloudevents.Event, subscription Subscriptio
 	var deadLetterURL *url.URL
 	if subscription.Delivery != nil && subscription.Delivery.DeadLetterSink != nil && !subscription.Delivery.DeadLetterSink.URI.IsEmpty() {
```
Comment (lionelvillard) [inline pkg/dispatcher/dispatcher/retry.go:61]: you may want to use `pkg/resolver/URIResolver` to support object reference

Comment (travis-minke-sap) [reply pkg/dispatcher/dispatcher/retry.go:61]: 👍  Not a part of original scope (existing code borrowed from eventing-contrib) but would be great to harden - thanks for suggesting!

**Comment by travis-minke-sap** [comment]
Hi @sbcd90 - Thanks for the PR and for looking into the proposed approach (detecting the final retry attempt)!

After going through the PR I see that you've been forced to copy the `slock/goresilience` retry logic out into a new function so that it could be modified to update the context in order to track the number of attempts.  Obviously that's not a great approach since we'd want to keep that copy of the implementation up to date with newer versions of the `slock/goresilience` library.  I don't see another way, however, to be able to track the state across the retry calls given the limited nature of what the library exposes.

So... instead, I think we'd prefer a simple solution like the one below where we just make an **additional** send request if the retries fail out with an error.  It's not fancy, but is easy to understand/maintain and is effective.

I've not tested this solution as of yet - only mocked it up in code.  And there are a couple **TODOs** there to look into the resolver library that @lionelvillard mentioned above.  If you're willing to continue with it (test it out, write/improve the unit tests, research the resolvers, etc) that would be awesome.  I realize most of the "new code" aspect (the fun part) has been eliminated, but it seems like you've done a great job of jumping into the code base and it would be great to get this delivered.  I'm open to your thoughts if I've mis-understood something or your have another idea ; ). Thanks again for contributing!


Proposed solution...
```
	var deadLetterURL *url.URL
	if subscription.Delivery != nil && subscription.Delivery.DeadLetterSink != nil && !subscription.Delivery.DeadLetterSink.URI.IsEmpty() { // TODO - use pkg/resolver/URIResolver?
		deadLetterURL = subscription.Delivery.DeadLetterSink.URI.URL() // TODO - use pkg/resolver/URIResolver?
	}

	// Attempt To Dispatch The CloudEvent Message Via Knative Message Dispatcher With Retry Wrapper
	err := retryRunner.Run(context.Background(), func(ctx context.Context) error {
		err := d.messageDispatcher.DispatchMessage(ctx, message, nil, destinationURL, replyURL, nil)
		return d.logResponse(err)
	})

	// Retry Attempts Failed
	if err != nil {

		// If The DeadLetterSink Is Configured
		if deadLetterURL != nil {

			// Then Make One Final Attempt With It & Return Result
			logger.Info("Making One Final Attempt With DeadLetterSink Configured")
			err := d.messageDispatcher.DispatchMessage(context.Background(), message, nil, destinationURL, replyURL, deadLetterURL)
			if err != nil {
				logger.Error("Failed To Send Final Attempt With DeadLetterSink", zap.Error(err))
			}
			return err

		} else {

			// Otherwise Return Original Retry Failure (DeadLetter Sink Not Configured)
			logger.Error("Failed To Send After Configured Number Of Retries - DeadLetterSink Not Configured", zap.Error(err))
			return err
		}
	}

	// Return Success
	return nil
```

**Comment by sbcd90** [comment]
@travis-minke-sap @lionelvillard thanks a lot for reviewing. I'll continue working on this PR as per your suggestions.

**Comment by travis-minke-sap** [comment]
/close

Closing PR as work has been done by #34 and will be further refined to use the new-ish knative DispatchWithRetry() functionality in the future.

**Pull request closed without merging**